### PR TITLE
GH-37796: [C++][Acero] Fix race condition caused by straggling input in the as-of-join node

### PR DESCRIPTION
**Pull request by JerAguilon**
### What changes are included in this PR?

While asofjoining some large parquet datasets with many row groups, I ran into a deadlock that I described here: https://github.com/apache/arrow/issues/37796. Copy pasting below for convenience:

1. The left hand side of the asofjoin completes and is matched with the right hand tables, so `InputFinished` proceeds as [expected](https://github.com/apache/arrow/blob/2455bc07e09cd5341d1fabdb293afbd07682f0b2/cpp/src/arrow/acero/asof_join_node.cc#L1323). So far so good
2. The right hand table(s) of the join are a huge dataset scan. They're still streaming and can legally still call `AsofJoinNode::InputReceived` all they want ([doc ref](https://arrow.apache.org/docs/cpp/api/acero.html#_CPPv4N5arrow5acero8ExecNode13InputReceivedEP8ExecNode9ExecBatch))
3. Each input batch is blindly pushed to the `InputState`s, which in turn defer to `BackpressureHandler`s to decide whether to pause inputs. ([code pointer](https://github.com/apache/arrow/blob/2455bc07e09cd5341d1fabdb293afbd07682f0b2/cpp/src/arrow/acero/asof_join_node.cc#L1689))
4. If enough batches come in right after `EndFromProcessThread` is called, then we might exceed the [high_threshold](https://github.com/apache/arrow/blob/2455bc07e09cd5341d1fabdb293afbd07682f0b2/cpp/src/arrow/acero/asof_join_node.cc#L575) and tell the input node to pause via the [BackpressureController](https://github.com/apache/arrow/blob/2455bc07e09cd5341d1fabdb293afbd07682f0b2/cpp/src/arrow/acero/asof_join_node.cc#L540)
5. At this point, the process thread has stopped for the asofjoiner, so the right hand table(s) won't be dequeue'd, meaning `BackpressureController::Resume()` will never be called. This causes a [deadlock](https://arrow.apache.org/docs/cpp/api/acero.html#_CPPv4N5arrow5acero19BackpressureControl5PauseEv)

TLDR this is caused by a straggling input node being paused due to backpressure _after_ the process thread has ended. And since every `PauseInput` needs a corresponding `ResumeInput` to exit gracefully, we deadlock.

Turns out this is fairly easy to reproduce with small tables, if you make a slow input node composed of 1-row record batches with a synthetic delay. 

My solution is to:

1. Create a `ForceShutdown` hook that puts the input nodes in a resumed state, and for good measure we call `StopProducing`
2. Also for good measure, if nodes come after the process thread exits, we short circuit and return OK. This is because `InputReceived` can be called an arbitrary number of times after `StopProducing`, so it makes sense to not enqueue useless batches.

### Are these changes tested?

Yes, I added a delay to the batches of one of the already-existing asofjoin backpressure tests. Checkout out `main`, we get a timeout failure. With my changes, it passes.

I considered a more deterministic test, but I struggled to create callbacks in a way that wasn't invasive to the Asof implementation. The idea of using delays was inspired by things I saw in `source_node_test.cc`

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
3. Serve as another way to document the expected behavior of the code

### Are there any user-facing changes?

No

* Closes: #37796

### CODE REVIEW

File: cpp/src/arrow/acero/asof_join_node.cc [hunk 1]
```diff
@@ -579,7 +581,16 @@ class BackpressureHandler {
     }
   }
 
+  Status ForceShutdown() {
+    // It may be unintuitive to call Resume() here, but this is to avoid a deadlock.
+    // Since acero's executor won't terminate if any one node is paused, we need to
+    // force resume the node before stopping production.
+    backpressure_control_->Resume();
```
Comment (JerAguilon) [inline cpp/src/arrow/acero/asof_join_node.cc:588]: Perhaps one thing to clarify is whether `ResumeInput` behaves idempotently? I.e., is it OK to always call resume, even though only some inputs hit this `PauseInput` race condition?

My perusal of `source_node.cc` tells me this is OK, but LMK if this is a poor assumption to make.

File: cpp/src/arrow/acero/asof_join_node.cc [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 #include <atomic>
 #include <condition_variable>
+#include <iostream>
```
Comment (JerAguilon) [inline cpp/src/arrow/acero/asof_join_node.cc:22]: Will remove

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 3]
```diff
@@ -1474,9 +1477,16 @@ void TestBackpressure(BatchesMaker maker, int num_batches, int batch_size) {
   std::vector<Declaration::Input> bp_decls;
   for (size_t i = 0; i < source_configs.size(); i++) {
     const auto& config = source_configs[i];
-
-    src_decls.emplace_back("source",
-                           SourceNodeOptions(config.schema, GetGen(config.batches)));
+    if (config.is_delayed) {
```
Comment (westonpace) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1480]: I assume this new option triggers the deadlock on the unfixed code?

**Review by westonpace — APPROVED** [review]
This seems like a good idea to me.

@icexelloss @rtpsw do either of you want to take a look?

File: cpp/src/arrow/acero/asof_join_node.cc [hunk 4]
```diff
@@ -579,7 +581,16 @@ class BackpressureHandler {
     }
   }
 
+  Status ForceShutdown() {
+    // It may be unintuitive to call Resume() here, but this is to avoid a deadlock.
+    // Since acero's executor won't terminate if any one node is paused, we need to
+    // force resume the node before stopping production.
+    backpressure_control_->Resume();
+    return input_->StopProducing();
```
[the remark this answers is not in the document]
Comment (JerAguilon) [reply cpp/src/arrow/acero/asof_join_node.cc:589]: Yep.

As an aside, I feel like a more invasive change could fix this issue in the general case. If a node (in this example asof join) has:

* Called `output->InputFinished()`  AND
* Called `output_->InputReceived` for however many record batches it advertised on `InputFinished`

We should be able to shut down execution, even if the node's inputs:

* are paused or
* not done streaming
* haven't called  `InputFinished`

But I think this is a more invasive change to `exec_plan.h` and might have some hairy issues that I'm not thinking of.

**Comment by icexelloss** [comment]
This looks reasonable to me. Free feel to merge.

File: cpp/src/arrow/acero/asof_join_node.cc [hunk 5]
```diff
@@ -1679,6 +1704,15 @@ class AsofJoinNode : public ExecNode {
   const Ordering& ordering() const override { return ordering_; }
 
   Status InputReceived(ExecNode* input, ExecBatch batch) override {
+    // InputReceived may be called after execution was finished. Pushing it to the
+    // InputState may cause the BackPressureController to pause the input, causing a
+    // deadlock
```
Comment (bkietz) [inline cpp/src/arrow/acero/asof_join_node.cc:1709]:
```suggestion
    // InputReceived may be called after execution was finished. Pushing it to the
    // InputState is unnecessary since we're done (and anyway may cause the
    // BackPressureController to pause the input, causing a deadlock), so drop it.
```

Do we still deadlock *with* this short circuit but *without* ForceShutdown etc?

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/acero/asof_join_node.cc (hunk 5 above)
Comment (JerAguilon) [reply cpp/src/arrow/acero/asof_join_node.cc:1706]: Yes, the `forceShutdown` is still necessary. there's nothing stopping this order of events:

1. We receive enough data to finish the as of join.
2. Right before we finish processing and shut down the worker thread, lots of batches come in from input A. Input A pauses
3. We shut down the thread, and input A can't be unpaused

But your comment change suggestions  sound good to me

**Comment by JerAguilon** [comment]
Clarifying comment for @bkietz added. Ready for more thoughts

**Review by bkietz — APPROVED** [review]

**Comment by JerAguilon** [comment]
Forgive the ignorance - first time making a PR on arrow. There's no further action needed from me to merge, correct?

**Comment by bkietz** [comment]
Could you rebase to pick up the fix https://github.com/apache/arrow/pull/37867 ? I think CI should be green after that

**Comment by JerAguilon** [comment]
Sadly there are some unrelated failures: `TestS3FS.GetFileInfoGeneratorStress` and `arrow-threading-utility-test`

**Comment by bkietz** [comment]
CI failures seem unrelated. I'll merge. Thanks for working on this!

**Comment by raulcd** [comment]
Closes: https://github.com/apache/arrow/issues/37796
